### DISCUSSION TITLE
Proposal: Enable use Span in generic delegates

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#1847] [Language Ideas]
Enable use Span in generic delegates.

For example as input parameter `Func<Span<T>, T, int>` and as output parameter `Func<T, T, Span<T>>`.

**Comment by HaloFour** [comment]
Why is it not sufficient to be able to define your own delegate types that accept/return `Span<T>`?  It's not like you can take advantage of generic variance in these situations since `Span<T>` is a value type.

**Comment by AlexRadch** [comment]
@HaloFour Any `struct` is a value type. `Span` is `ref struct`.

**Comment by HaloFour** [comment]
@AlexRadch 

Yes, `ref struct`s are just `structs` with imposed limitations.

The CLR forbids `struct` generic type arguments to participate in variance, so you can't assign a `Func<int>` to a `Func<object>`, whereas you can assign a `Func<string>` to a `Func<object>`.  So, with variance not possible, I don't see what the benefit would be to lift the imposed restriction that `ref struct`s can't be generic type arguments where the generic type parameter is defined on a `delegate`.  You're always free to define your own `delegate`, which is the appropriate route when the generic delegate families of `Func<...>` or `Action<...>` don't meet your need.

**Comment by AlexRadch** [comment]
`Span` can not be assigned to or from object, so `Func<Span>` can not be assigned to or from `Func<object>`. But it is not means that  `Func<Span>` should be disabled as now.

**Comment by BladeWise** [comment]
@AlexRadch No, you can't. 
``` C#
            Func<int> getInt = () => 1;
            Func<object> getObj = getInt;
```
does not compile. Covariance is available either through interface or type inheritance, if I am not mistaken.

**Comment by svick** [comment]
@AlexRadch So what is the argument for enabling it? As far as I can tell, you haven't given any reason as for why this work should be done.

**Comment by AlexRadch** [comment]
@Svick What argument for disabling it?

**Comment by AlexRadch** [comment]
For example why I should define separated delegate,  where I can use `Func`?
```cs
    public delegate int IndexOfFunc<T>(Span<T> span, T value);

    public class MemoryExtensionsTests_IndexOfEqual<T>
    {
        public MemoryExtensionsTests_IndexOfEqual(IndexOfFunc<T> indexOf)
    }
```
Just use `Func` is smart
```cs
    public class MemoryExtensionsTests_IndexOfEqual<T>
    {
        public MemoryExtensionsTests_IndexOfEqual(Func<Span<T>, T, int> indexOf)
    }
```

**Comment by svick** [comment]
@AlexRadch That's not how this works. The current state is that it's disabled. You want to enable it, in other words, you're proposing a new language feature. And since features are a lot of work, they require justification. If you can'y provide that, I think the chance that will be accepted to the language is incredibly  low.
